servicetalk-loadbalancer/src/main/java/io/servicetalk/loadbalancer/P2CSelector.java

```diff
@@ -78,58 +87,73 @@ private Single<C> p2c(int size, List<Host<ResolvedAddress, C>> hosts, Random ran
             if (i2 >= i1) {
                 ++i2;
             }
+
             Host<ResolvedAddress, C> t1 = hosts.get(i1);
             Host<ResolvedAddress, C> t2 = hosts.get(i2);
-            // Make t1 the preferred host by score to make the logic below a bit cleaner.
-            if (t1.score() < t2.score()) {
-                Host<ResolvedAddress, C> tmp = t1;
-                t1 = t2;
-                t2 = tmp;
-            }
-
-            // Attempt to get a connection from t1 first since it's 'better'. If we can't, then try t2.
-            Single<C> result = selectFromHost(t1, selector, forceNewConnectionAndReserve, contextMap);
-            if (result != null) {
-                return result;
+            final boolean t1Healthy = t1.isHealthy();
+            final boolean t2Healthy = t2.isHealthy();
+            // Priority of selection: health > score > failOpen
+            // Only if both hosts are healthy do we consider score.
+            if (t1Healthy && t2Healthy) {
+                // both are healthy. Select based on score, using t1 if equal.
+                if (t1.score() < t2.score()) {
+                    Host<ResolvedAddress, C> tmp = t1;
+                    t1 = t2;
+                    t2 = tmp;
+                }
+                Single<C> result = selectFromHost(
+                        t1, selector, forceNewConnectionAndReserve, contextMap);
+                // We didn't get a connection from the first host: maybe it is inactive
+                // and we couldn't reserve a connection. Try the second host.
+                if (result == null) {
+                    result = selectFromHost(t2, selector, forceNewConnectionAndReserve, contextMap);
+                }
+                // If we have a connection we're good to go. Otherwise fall through for another round.
+                // Since we didn't get a connection from either of them there is no reason to think they'll
+                // yield a connection if we make them the fallback.
+                if (result != null) {
+                    return result;
+                }
+            } else if (t2Healthy) {
+                Single<C> result = selectFromHost(t2, selector, forceNewConnectionAndReserve, contextMap);
+                if (result != null) {
+                    return result;
+                }
+            } else if (t1Healthy) {
+                Single<C> result = selectFromHost(t1, selector, forceNewConnectionAndReserve, contextMap);
+                if (result != null) {
+                    return result;
+                }
+            } else if (failOpen && failOpenHost == null) {
+                // Both are unhealthy. If one of them can make new connections then it can be a backup.
+                if (t1.canMakeNewConnections()) {
+                    failOpenHost = t1;
+                } else if (t2.canMakeNewConnections()) {
+                    failOpenHost = t2;
+                }
```

Comment:
I think there is a good case to be made that if we want to fail open we shouldn't even worry about whether it's active or not: that is arguably not a show stopper if you're willing to  yo-lo it. Opinions appreciated.